apps/src/makerlab/BoardController.js

```diff
@@ -0,0 +1,197 @@
+var five = require('johnny-five');
+var ChromeSerialPort = require('chrome-serialport');
+var PlaygroundIO = require('playground-io');
+require("babelify/polyfill"); // required for Promises in IE / Phantom
+
+/** @const {string} */
+var CHROME_APP_ID = 'himpmjbkjeenflliphlaaeggkkanoglo';
+
+var BoardController = module.exports = function () {
+  ChromeSerialPort.extensionId = CHROME_APP_ID;
+
+  /** @private {five.Board} */
+  this.board_ = null;
+
+  this.prewiredComponents = null;
+};
+
+BoardController.prototype.connectAndInitialize = function (codegen, interpreter) {
+  return this.ensureBoardConnected()
+      .then(this.installComponentsOnInterpreter.bind(this, codegen, interpreter))
+      .catch(function (reason) {
+        console.log("Board initialization failed:");
+        console.log(reason);
+      });
+};
+
+/**
+ * Connects to board if not already connected.
+ */
+BoardController.prototype.ensureBoardConnected = function () {
+  return new Promise(function (resolve, reject) {
+    ChromeSerialPort.isInstalled(function (error) {
+      if (error) {
+        reject(error);
+        return;
+      }
+
+      if (this.board_) {
+        resolve();
+        return;
+      }
+
+      connect()
+          .then(function (board) {
+            this.board_ = board;
+          }.bind(this))
+          .then(resolve)
+          .catch(reject);
```

Comment:
Feels like this we should be able to simplify this, but I'm not exactly sure how.

One thing we could do is to move the .then(resolve) into the previous function, i.e.
```
connect()
  .then(function (board) {
    this.board_ = board;
    resolve();
  })
  .catch(reject)